### ISSUE TITLE
Descent II (Win95 version) Crashes on level load 50% of the time

### ISSUE THREAD

**Issue by OpenRift412** [narzoul/DDrawCompat#199]
As the title suggests. It works fine in the intro and menus, but once it tries to load the level, it's like a 50/50 chance that it hits me with an error like this:
![image](https://user-images.githubusercontent.com/29780788/217150090-60e59c2a-4de2-439d-90ea-595bdba77d21.png)

Any idea what might be causing this?

**Comment by BEENNath58** [comment]
This is an infamous heap error. Install the EmulateHeap shim using Microsoft Application Compatibility Toolkit

**Comment by OpenRift412** [comment]
> This is an infamous heap error. Install the EmulateHeap shim using Microsoft Application Compatibility Toolkit

Is there any way to make this fix portable?

**Comment by narzoul** [comment]
Well, it's working better for you than for me then. I get a crash 100% of the time when starting a new game. EmulateHeap doesn't help, but I'm also on Windows 11 22H2.

I tried the 1.2 patch, which just adds another issue (the game doesn't unlock a surface properly and then complains about it and exits), but after bypassing that issue, it just crashes anyway.

As far as I can tell, none of the crashes are related to DirectDraw, so there's not much I can do about it.

**Comment by BEENNath58** [comment]
I thought I should declare: On Win11 22H2, EmulateHeap removes the sound primarily. Also, I faced noticeable pauses in menus, probably the game was on the verge to crash. So yes, that's a no-go for EmulateHeap.

**Comment by narzoul** [comment]
I'm closing this, as I don't think I'll have the time for individual heap patches for games anytime soon.

**Comment by BEENNath58** [comment]
@OpenRift412 

I tried the latest EmulateHeap patch from crazyc
Without DDrawCompat the game crashed 0/12 times
With DDrawCompat it crashes 8/12 times

I am not sure if DDC is stressing the heap compartment for the game

**Comment by OpenRift412** [comment]
> [@OpenRift412](https://github.com/OpenRift412)
> 
> I tried the latest EmulateHeap patch from crazyc Without DDrawCompat the game crashed 0/12 times With DDrawCompat it crashes 8/12 times
> 
> I am not sure if DDC is stressing the heap compartment for the game

I assume the EmulateHeap thing is part of DxWnd? Could you send me your DxWnd config?

**Comment by BEENNath58** [comment]
> I assume the EmulateHeap thing is part of DxWnd? Could you send me your DxWnd config?

Download the latest DxWnd and use this RC:
https://sourceforge.net/p/dxwnd/discussion/general/thread/c284b165c6/?page=2#91d7

My config is simple, enable Libs/Emulate Win9x Heap and since I prefer CD audio, I used CDaudip/Play from CD

**Comment by OpenRift412** [comment]
> > I assume the EmulateHeap thing is part of DxWnd? Could you send me your DxWnd config?
> 
> Download the latest DxWnd and use this RC: https://sourceforge.net/p/dxwnd/discussion/general/thread/c284b165c6/?page=2#91d7
> 
> My config is simple, enable Libs/Emulate Win9x Heap and since I prefer CD audio, I used CDaudip/Play from CD (maybe you should Reopen the issue so that I don't have to search it to reply everytime)

Yeah I got it working, only problem is I can't seem to get it working in proxy mode.

**Comment by BEENNath58** [comment]
> I can't seem to get it working in proxy mode.

I don't think the heap module will be working in proxy mode as of now because it uses a different injection tool. I can still make a request to crazyc if that's possible, but there are other things to solve in this heap module first, and to make it correctly portable (the previous build only worked on Windows 10 and first Windows 11 release)

**Comment by ArhumMK** [comment]
Is it possible to implement EmulateHeap into ddrawcompat? I saw that DxWnd has something like it now though no idea how to use it, Last Bronx is so close to running perfectly, all that's left is making a portable EmulateHeap shim and figuring out the video tabbing out/crashing issue.

**Comment by narzoul** [comment]
No, I'm not planning on such a thing. I don't believe in such solutions in the first place, these games need real patches to remain stable long term. And even if I wanted to do it, EmulateHeap is an unchanged copy of the Win95 heap management functions from what I understand, which is not exactly open source. I don't want any legal troubles.

**Comment by BEENNath58** [comment]
@ArhumMK Last Bronx works fine now with DxWnd's emulateheap methods. A portable solution will still be difficult I reckon, even if Narzoul gets to make a complete open-source release.

For any other bugs I didn't encounter in that game, you could share back
